scalasql/query/src-3/TableMacro.scala

```diff
@@ -0,0 +1,177 @@
+package scalasql.query
+
+import scalasql.core.{DialectTypeMappers, Expr => SqlExpr, Queryable, Sc, TypeMapper}
+import scala.compiletime.summonInline
+import scala.quoted.*
+
+object TableMacros {
+  def applyImpl[V[_[_]] <: Product](using Quotes, Type[V]): Expr[Table.Metadata[V]] = {
+    import quotes.reflect.*
+
+    val caseClassType = TypeRepr.of[V]
+    val constructor = caseClassType.typeSymbol.primaryConstructor
+    val constructorTypeParams = constructor.paramSymss(0)
+    val constructorValueParams = constructor.paramSymss(1)
+
+    def paramType(param: Symbol): TypeRepr = caseClassType.memberType(param)
+
+    def isTypeParamType(param: Symbol): Boolean =
+      paramType(param).typeSymbol.toString != constructorTypeParams.head.toString
+
+    def subParam(paramInfo: TypeRepr, tpe: TypeRepr): TypeRepr =
+      paramInfo.substituteTypes(List(constructorTypeParams.head), List(tpe))
+
+    val queryables = '{ (dialect: DialectTypeMappers, n: Int) =>
+      {
+        @annotation.nowarn("msg=unused")
+        given d: DialectTypeMappers = dialect
+
+        ${
+          Expr.ofList(constructorValueParams.map { param =>
+            val paramTpe = paramType(param)
+            val tpe = subParam(paramTpe, TypeRepr.of[Sc])
+            val tpe2 = subParam(paramTpe, TypeRepr.of[SqlExpr])
+            (tpe.asType, tpe2.asType) match {
+              case ('[t], '[t2]) => '{ summonInline[Queryable.Row[t2, t]] }
+            }
+          })
+        }.apply(n)
+      }
+    }
+
+    val walkLabels0 = '{ () =>
+      ${
+        Expr.ofList(constructorValueParams.map { param =>
+          if (isTypeParamType(param))
+            paramType(param) match {
+              case AppliedType(tpeCtor, _) =>
+                tpeCtor.asType match {
+                  case '[
+                      type t[_[_]]; t] =>
+                    '{ summonInline[Table.ImplicitMetadata[t]].value.walkLabels0() }
+                }
+            }
```

Comment:
This whole block appears twice; could it be moved into a named helper method, with the final `.walkLabels0()` or `.vExpr(tableRef, dialect)` passed as a lambda or something to configure it?

Reply:
Yeah good idea, updated to add a helper method for this